platform/lang-impl/src/com/intellij/build/ExecutionNode.java

```diff
@@ -215,16 +238,21 @@ public void setFilter(@Nullable Predicate<ExecutionNode> filter) {
     for (ExecutionNode node : myChildrenList) {
       node.setFilter(myFilter);
     }
-    cleanUpCache();
+    if (filter == null) {
+      myVisibleChildrenList = null;
+    }
+    else {
+      if (myVisibleChildrenList == null) {
+        myVisibleChildrenList = Collections.synchronizedList(new ArrayList<>());
+      }
+      maybeReapplyFilter();
+    }
   }
 
-  public void setVisible(boolean visible) {
-    if (myVisible != visible) {
-      myVisible = visible;
-      SimpleNode parent = getParent();
-      if (parent instanceof CachingSimpleNode) {
-        ((CachingSimpleNode)parent).cleanUpCache();
-      }
+  private void maybeReapplyFilter() {
+    if (myVisibleChildrenList != null) {
+      myVisibleChildrenList.clear();
+      myVisibleChildrenList.addAll(myChildrenList.stream().filter(it -> nodeIsVisible(it)).collect(Collectors.toList()));
```

Comment:
Using `forEachOrdered` may save you building a large intermediate list (e.g., for thousands of nodes):
`myChildrenList.stream().filter(it -> nodeIsVisible(it)).forEachOrdered(visibleChildrenList::add)`

Reply:
Done